README.md
ansible-role-influxdb
=========

Ansible role to install/configure InfluxDB from InfluxDB binaries.  
Available on [Ansible Galaxy](https://galaxy.ansible.com/rockandska/influxdb)

**Ansible Galaxy:**  
![Galaxy Score](https://img.shields.io/ansible/quality/38392.svg)

**Travis Build :**  
[![Build Status](https://travis-ci.com/rockandska/ansible-role-influxdb.png?branch=master)](https://travis-ci.com/rockandska/ansible-role-influxdb) 

Requirements on remote hosts
------------

#### All distro

- logrotate
- python requests >= 1.0.0 ( if using users,databases,retention_policies management provide by this role )
- python influxdb >= 0.9.0  ( if using users,databases,retention_policies management provide by this role )

#### Debian / Ubuntu

- apt-transport-https
- gpg-agent
- ca-certificates

#### CentOS / RedHat
- gnupg2

Role Variables
--------------

Defaults variables are inside `defaults/main.yml`
```yaml
---
###########
# Install #
###########
influxdb_version: 1.7.4

influxdb_rpm_url: "https://dl.influxdata.com/influxdb/releases/influxdb-{{ influxdb_version }}.x86_64.rpm"
influxdb_rpm_sha256: "93caa49159e76b3338abc191c85d46d134ac67697a0c7f874c937baa38dd6a5e"
influxdb_rpm_dest_path: "/usr/src"

influxdb_deb_url: "https://dl.influxdata.com/influxdb/releases/influxdb_{{ influxdb_version }}_amd64.deb"
influxdb_deb_sha256: "5488bd8889de7d45aa09cdccc845156e2d59442871ad6899d64d797e42f2888b"
influxdb_deb_dest_path: "/usr/src"

#################
# Custom Config #
#################
influxdb_vars_files: []

influxdb_config_tpl: etc/influxdb/influxdb.conf.j2
influxdb_config: {}

influxdb_systemd_override_tpl: etc/systemd/system/influxdb.service.d/override.conf.j2
influxdb_systemd_override: {}

influxdb_custom_logrotate_tpl: etc/logrotate.d/influxdb.j2
influxdb_custom_logrotate:

############
# Api user #
############
influxdb_management_hostname:
influxdb_management_port:
influxdb_management_validate_certs:
influxdb_management_proxies:
influxdb_management_login_username:
influxdb_management_login_password:
influxdb_management_retries:
influxdb_management_ssl:
influxdb_management_timeout:
influxdb_management_udp_port:
influxdb_management_use_udp:

#########
# Users #
#########
influxdb_users_to_create: []
influxdb_users_to_delete: []

#############
# Databases #
#############
influxdb_databases_to_create: []
influxdb_databases_to_delete: []

############
# Policies #
############
influxdb_retention_policies_to_create: []

#########
# Debug #
#########
influxdb_hide_log: true
```

#### Details

- `influxdb_version`

  - should be a influxdb version available on https://dl.influxdata.com/influxdb/releases
  - old versions are not shown on download page but are available from direct links

- `influxdb_rpm_url`

  - url of the rpm package

- `influxdb_rpm_sha256`

  - sha256 of the rpm package to verify its integrity
  - Available on InfluxDB download page for the last release

- `influxdb_rpm_dest_path`

  - path where to download the rpm package

- `influxdb_deb_url`

  - url of the deb package

- `influxdb_deb_sha256`

  - sha256 of the deb package to verify its integrity
  - Available on InfluxDB download page for the last release

- `influxdb_deb_dest_path`

  - path where to download the deb package

- `influxdb_vars_files`

  - list of vars files used to override defaults variables if needed
  - if using relative path, put those files next to your playbook in `vars` directory
  - example:
    ```yaml
    influxdb_vars_files:
      - settings.yml
    ```

- `influxdb_config_tpl`

  - path to the influxdb config template
  - if you want to use your own template
    - add your template next to your playbook in a `templates` directory
    - use a different path than the default one

- `influxdb_config`

  - a dict representing the custom influxdb config to apply

  - put specials keys between double quotes (example: "[meta]")

  - If using custom config, some minimal sections need to be apply due to [influxdb/#12140](https://github.com/influxdata/influxdb/issues/12140) and are shown in the example bellow

  - examples:
    ```yaml
    influxdb_config:
    # Global config section as 'key: value'
      reporting-disabled: true
    # Minimal required sections as 'section: { key: value, key: value }'
      "[meta]":
        dir: /var/lib/influxdb/meta
      "[data]":
        dir: /var/lib/influxdb/data
        wal-dir: /var/lib/influxdb/wal
    # Other sections in format as 'section: { key: value, key: value }'
      "[http]":
        auth-enabled: true
      "[[graphite]]":
        enable: true
        batch-size: 5000
    ```

- `influxdb_systemd_override_tpl`

  - path to the influxdb systemd override template
  - if you want to use your own template
    - add your template next to your playbook in a `templates` directory
    - use a different path than the default one

- `influxdb_systemd_override`

  - a dict representing the systemd override config
  - the first level is used for the ini section
  - the second level is used for the key / value 
  - example:
    ```yaml
    influxdb_systemd_override:
      Service:
        LimitNOFILE: 30000
    # Will result into the systemd override file as:
    # [Service]
    # LimitNOFILE=30000
    ```

- `influxdb_custom_logrotate_tpl`

  - path to the influxdb custom logrotate template
  - if you want to use your own template
    - add your template next to your playbook in a `templates` directory
    - use a different path than the default one

- `influxdb_custom_logrotate`

  - a multiline string with the logrotate options for influxdb logs

  - will erase the default config

  - **/!\ Be aware that if you replace the default logrotate config by a custom one, the configuration applied will persist even if you unset this variable**

  - example:
    ```yaml
    influxdb_custom_logrotate: |
      weekly
      missingok
      rotate 40
      compress
      notifempty
    # Will result into the logrotate config file as:
    # /var/log/influxdb/*.log {
    #   weekly
    #   missingok
    #   rotate 40
    #   compress
    #   notifempty
    # }
    ```

- `influxdb_management_hostname`
  - default from influxdb_{user,policy,retention_policy} module value
  - hostname for API calls

- `influxdb_management_port`
  - default from influxdb_{user,policy,retention_policy} module value
  - http port for API calls

- `influxdb_management_validate_certs`
  - default from influxdb_{user,policy,retention_policy} module value
  - do we need to validate the SSL certificate for API calls

- `influxdb_management_proxies`
  - default from influxdb_{user,policy,retention_policy} module value
  - HTTP(S) proxy to use for API calls

- `influxdb_management_login_username`
  - username to use for API calls
  - **mandatory if `influxdb_config['[http]']['auth-enabled']` is set**
  - This user will be create only once, and need to be a **none already existing user**
  - If you want to change the user afterward its creation, use an already existing `admin` user.

- `influxdb_management_login_password`

  - password to use for API calls
  - **mandatory if `influxdb_config['[http]']['auth-enabled']` is set**
  - If you want to change the password afterward the `admin` user creation:
    - Change the password through the CLI before changing it here
    - Alternatively, you could change it into `influxdb_users_to_create`, but be aware that it will throw an error if you have remaining users who need to be checked but should works if you relaunch your playbook a second time.

- `influxdb_management_retries`

  - default from influxdb_{user,policy,retention_policy} module value
  - number of retries for API calls

- `influxdb_management_ssl`

  - default from influxdb_{user,policy,retention_policy} module value
  - use SSL for API calls

- `influxdb_management_timeout`
  - default from influxdb_{user,policy,retention_policy} module value
  - timeout for API calls

- `influxdb_management_udp_port`
  - default from influxdb_{user,policy,retention_policy} module value
  - UDP port to use for API calls

- `influxdb_management_use_udp`
  - default from influxdb_{user,policy,retention_policy} module value
  - do we need to use UDP for API calls

- `influxdb_users_to_create`

  - list of dict for users creation

  - refer to [ansible doc](https://docs.ansible.com/ansible/latest/modules/influxdb_user_module.html) for mandatory options and version compatibility

  - **grant privileges are provided by a custom library provided with this role and update from this [PR](https://github.com/ansible/ansible/pull/46216)**

  - **the absence of `grants` key will remove all `GRANTS` for the user specified**

  - Example:
    ```yaml
    influxdb_users_to_create:
      - user_name: test
        user_password: test
        grants:
          - database: db_test
            privilege: READ  
      - user_name: test2
        user_password: test2
        admin: true  
    ```

- `influxdb_users_to_delete`
  - list of users to delete
  - Example:
    ```yaml
    influxdb_users_to_delete:
      - test
    ```

- `influxdb_retention_policies_to_create`
  - list of retention policies to create
  - refer to [ansible doc](https://docs.ansible.com/ansible/latest/modules/influxdb_retention_policy_module.html) for mandatory options and version compatibility
  - example:

    ```yaml
    influxdb_retention_policies_to_create:
      - database_name: db_test
        policy_name: test
        duration: 1h
        replication: 1
    ```

- `influxdb_databases_to_create`

    - list of databases to create
    - examples:

  ```yaml
  influxdb_databases_to_create:
    - db_test
  ```

- `influxdb_databases_to_delete`


    - list of databases to delete
    
    - examples:

  ```yaml
  influxdb_databases_to_delete:
    - db_test
  ```

- `influxdb_retention_policies_to_create`


    - list of dict retention policies parameters
    
    - **deletion of policies are not supported yet**
    
    - examples:

  ```yaml
  influxdb_retention_policies_to_create:
    - database_name: db_test
      policy_name: test
      duration: 1h
      replication: 1  
  ```

- `influxdb_hide_log`

  - default: true
  - don't show the log for api calls to avoid leak of sensible informations
  - set it to false for debug

Example Playbook
----------------

```yaml
- hosts: influxdb
  roles:
    - rockandska.influxdb
```

Local Testing
-------------

The preferred way of locally testing the role is to use Docker and [molecule](https://github.com/metacloud/molecule) (v2.x). You will have to install Docker on your system. See "Get started" for a Docker package suitable to for your system.
We are using tox to simplify process of testing on multiple ansible versions. To install tox execute:
```sh
$ sudo pip install tox
# or
$ pip install --user tox
```

To run tests on all ansible versions (WARNING: this can take some time)
```sh
$ tox
```
To run a custom molecule command on custom environment with only default test scenario:
```sh
$ tox -e py27-ansible25 -- molecule test -s default
```
For more information about molecule go to their [docs](http://molecule.readthedocs.io/en/latest/).

If you would like to run tests on remote docker host just specify `DOCKER_HOST` variable before running tox tests.

License
-------

BSD


CHANGELOG.md
# Change Log

## [0.0.3](https://github.com/rockandska/ansible-role-influxdb/tree/0.0.3) (2019-07-01)
**Merged pull requests:**

- Add options for management user creation [\#3](https://github.com/rockandska/ansible-role-influxdb/pull/3) ([rockandska](https://github.com/rockandska))
- Fix pre\_check syntax [\#2](https://github.com/rockandska/ansible-role-influxdb/pull/2) ([rockandska](https://github.com/rockandska))

## [0.0.2](https://github.com/rockandska/ansible-role-influxdb/tree/0.0.2) (2019-03-06)
**Merged pull requests:**

- Fix custom module name [\#1](https://github.com/rockandska/ansible-role-influxdb/pull/1) ([rockandska](https://github.com/rockandska))

## [0.0.1](https://github.com/rockandska/ansible-role-influxdb/tree/0.0.1) (2019-03-04)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

library/influxdb_user.py
#!/usr/bin/python

# Update from
# https://github.com/ansible/ansible/pull/46216

# (c) 2017, Vitaliy Zhhuta <zhhuta () gmail.com>
# insipred by Kamil Szczygiel <kamil.szczygiel () intel.com> influxdb_database module
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
---
module: influxdb_user
short_description: Manage InfluxDB users
description:
  - Manage InfluxDB users
version_added: 2.5
author: "Vitaliy Zhhuta (@zhhuta)"
requirements:
  - "python >= 2.6"
  - "influxdb >= 0.9"
options:
  user_name:
    description:
      - Name of the user.
    required: True
  user_password:
    description:
      - Password to be set for the user.
    required: false
  admin:
    description:
      - Whether the user should be in the admin role or not.
      - Since version 2.8, the role will also be updated.
    default: no
    type: bool
  state:
    description:
      - State of the user.
    choices: [ present, absent ]
    default: present
  grants:
    description:
      - Privileges to grant to this user. Takes a list of dicts containing the
        "database" and "privilege" keys.
    default: []
    version_added: 2.8
extends_documentation_fragment: influxdb
'''

EXAMPLES = '''
- name: Create a user on localhost using default login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t

- name: Create a user on localhost using custom login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"

- name: Create an admin user on a remote host using custom login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    admin: yes
    hostname: "{{ influxdb_hostname }}"
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"

- name: Create a user on localhost with privileges
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"
    grants:
      - database: 'collectd'
        privilege: 'WRITE'
      - database: 'graphite'
        privilege: 'READ'

- name: Destroy a user using custom login credentials
  influxdb_user:
    user_name: john
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"
    state: absent
'''

RETURN = '''
#only defaults
'''

import ansible.module_utils.urls
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_native
import ansible.module_utils.influxdb as influx


def find_user(module, client, user_name):
    user_result = None

    try:
        users = client.get_list_users()
        for user in users:
            if user['user'] == user_name:
                user_result = user
                break
    except (ansible.module_utils.urls.ConnectionError, influx.exceptions.InfluxDBClientError) as e:
        module.fail_json(msg=str(e))
    return user_result


def check_user_password(module, client, user_name, user_password):
    try:
        client.switch_user(user_name, user_password)
        client.get_list_users()
    except influx.exceptions.InfluxDBClientError as e:
        if e.code == 401:
            return False
    except ansible.module_utils.urls.ConnectionError as e:
        module.fail_json(msg=to_native(e))
    finally:
        # restore previous user
        client.switch_user(module.params['username'], module.params['password'])
    return True


def set_user_password(module, client, user_name, user_password):
    if not module.check_mode:
        try:
            client.set_user_password(user_name, user_password)
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=to_native(e))


def create_user(module, client, user_name, user_password, admin):
    if not module.check_mode:
        try:
            client.create_user(user_name, user_password, admin)
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=to_native(e))


def drop_user(module, client, user_name):
    if not module.check_mode:
        try:
            client.drop_user(user_name)
        except influx.exceptions.InfluxDBClientError as e:
            module.fail_json(msg=e.content)

    module.exit_json(changed=True)


def set_user_grants(module, client, user_name, grants):
    changed = False

    try:
        current_grants = client.get_list_privileges(user_name)
        # Fix privileges wording
        for i, v in enumerate(current_grants):
            if v['privilege'] == 'ALL PRIVILEGES':
                v['privilege'] = 'ALL'
                current_grants[i] = v
            elif v['privilege'] == 'NO PRIVILEGES':
                del(current_grants[i])


        # check if the current grants are included in the desired ones
        for current_grant in current_grants:
            if current_grant not in grants:
                if not module.check_mode:
                    client.revoke_privilege(current_grant['privilege'],
                                            current_grant['database'],
                                            user_name)
                changed = True

        # check if the desired grants are included in the current ones
        for grant in grants:
            if grant not in current_grants:
                if not module.check_mode:
                    client.grant_privilege(grant['privilege'],
                                           grant['database'],
                                           user_name)
                changed = True

    except influx.exceptions.InfluxDBClientError as e:
        module.fail_json(msg=e.content)

    return changed


def main():
    argument_spec = influx.InfluxDb.influxdb_argument_spec()
    argument_spec.update(
        state=dict(default='present', type='str', choices=['present', 'absent']),
        user_name=dict(required=True, type='str'),
        user_password=dict(required=False, type='str', no_log=True),
        admin=dict(default='False', type='bool'),
        grants=dict(default=[], type='list')
    )
    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True
    )

    state = module.params['state']
    user_name = module.params['user_name']
    user_password = module.params['user_password']
    admin = module.params['admin']
    grants = module.params['grants']
    influxdb = influx.InfluxDb(module)
    client = influxdb.connect_to_influxdb()
    user = find_user(module, client, user_name)

    changed = False

    if state == 'present':
        if user:
            if not check_user_password(module, client, user_name, user_password) and user_password is not None:
                set_user_password(module, client, user_name, user_password)
                changed = True

            try:
                if admin and not user['admin']:
                    client.grant_admin_privileges(user_name)
                    changed = True
                elif not admin and user['admin']:
                    client.revoke_admin_privileges(user_name)
                    changed = True
            except influx.exceptions.InfluxDBClientError as e:
                module.fail_json(msg=str(e))

        else:
            user_password = user_password or ''
            create_user(module, client, user_name, user_password, admin)
            changed = True

        if grants:
            if set_user_grants(module, client, user_name, grants):
                changed = True
        else:
            if set_user_grants(module, client, user_name, []):
                changed = True

        module.exit_json(changed=changed)

    if state == 'absent':
        if user:
            drop_user(module, client, user_name)
        else:
            module.exit_json(changed=False)


if __name__ == '__main__':
    main()



library/influxdb_user_custom.py
#!/usr/bin/python

# Update from
# https://github.com/ansible/ansible/pull/46216

# (c) 2017, Vitaliy Zhhuta <zhhuta () gmail.com>
# insipred by Kamil Szczygiel <kamil.szczygiel () intel.com> influxdb_database module
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = '''
---
module: influxdb_user
short_description: Manage InfluxDB users
description:
  - Manage InfluxDB users
version_added: 2.5
author: "Vitaliy Zhhuta (@zhhuta)"
requirements:
  - "python >= 2.6"
  - "influxdb >= 0.9"
options:
  user_name:
    description:
      - Name of the user.
    required: True
  user_password:
    description:
      - Password to be set for the user.
    required: false
  admin:
    description:
      - Whether the user should be in the admin role or not.
      - Since version 2.8, the role will also be updated.
    default: no
    type: bool
  state:
    description:
      - State of the user.
    choices: [ present, absent ]
    default: present
  grants:
    description:
      - Privileges to grant to this user. Takes a list of dicts containing the
        "database" and "privilege" keys.
    default: []
    version_added: 2.8
extends_documentation_fragment: influxdb
'''

EXAMPLES = '''
- name: Create a user on localhost using default login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t

- name: Create a user on localhost using custom login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"

- name: Create an admin user on a remote host using custom login credentials
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    admin: yes
    hostname: "{{ influxdb_hostname }}"
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"

- name: Create a user on localhost with privileges
  influxdb_user:
    user_name: john
    user_password: s3cr3t
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"
    grants:
      - database: 'collectd'
        privilege: 'WRITE'
      - database: 'graphite'
        privilege: 'READ'

- name: Destroy a user using custom login credentials
  influxdb_user:
    user_name: john
    login_username: "{{ influxdb_username }}"
    login_password: "{{ influxdb_password }}"
    state: absent
'''

RETURN = '''
#only defaults
'''

import ansible.module_utils.urls
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils._text import to_native
import ansible.module_utils.influxdb as influx


def find_user(module, client, user_name):
    user_result = None

    try:
        users = client.get_list_users()
        for user in users:
            if user['user'] == user_name:
                user_result = user
                break
    except (ansible.module_utils.urls.ConnectionError, influx.exceptions.InfluxDBClientError) as e:
        module.fail_json(msg=to_native(e))
    return user_result


def check_user_password(module, client, user_name, user_password):
    try:
        client.switch_user(user_name, user_password)
        client.get_list_users()
    except influx.exceptions.InfluxDBClientError as e:
        if e.code == 401:
            return False
    except ansible.module_utils.urls.ConnectionError as e:
        module.fail_json(msg=to_native(e))
    finally:
        # restore previous user
        client.switch_user(module.params['username'], module.params['password'])
    return True


def set_user_password(module, client, user_name, user_password):
    if not module.check_mode:
        try:
            client.set_user_password(user_name, user_password)
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=to_native(e))


def create_user(module, client, user_name, user_password, admin):
    if not module.check_mode:
        try:
            client.create_user(user_name, user_password, admin)
        except ansible.module_utils.urls.ConnectionError as e:
            module.fail_json(msg=to_native(e))


def drop_user(module, client, user_name):
    if not module.check_mode:
        try:
            client.drop_user(user_name)
        except influx.exceptions.InfluxDBClientError as e:
            module.fail_json(msg=e.content)

    module.exit_json(changed=True)


def set_user_grants(module, client, user_name, grants):
    changed = False

    try:
        current_grants = client.get_list_privileges(user_name)
        # Fix privileges wording
        for i, v in enumerate(current_grants):
            if v['privilege'] == 'ALL PRIVILEGES':
                v['privilege'] = 'ALL'
                current_grants[i] = v
            elif v['privilege'] == 'NO PRIVILEGES':
                del(current_grants[i])


        # check if the current grants are included in the desired ones
        for current_grant in current_grants:
            if current_grant not in grants:
                if not module.check_mode:
                    client.revoke_privilege(current_grant['privilege'],
                                            current_grant['database'],
                                            user_name)
                changed = True

        # check if the desired grants are included in the current ones
        for grant in grants:
            if grant not in current_grants:
                if not module.check_mode:
                    client.grant_privilege(grant['privilege'],
                                           grant['database'],
                                           user_name)
                changed = True

    except influx.exceptions.InfluxDBClientError as e:
        module.fail_json(msg=e.content)

    return changed


def main():
    argument_spec = influx.InfluxDb.influxdb_argument_spec()
    argument_spec.update(
        state=dict(default='present', type='str', choices=['present', 'absent']),
        user_name=dict(required=True, type='str'),
        user_password=dict(required=False, type='str', no_log=True),
        admin=dict(default='False', type='bool'),
        grants=dict(default=[], type='list')
    )
    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True
    )

    state = module.params['state']
    user_name = module.params['user_name']
    user_password = module.params['user_password']
    admin = module.params['admin']
    grants = module.params['grants']
    influxdb = influx.InfluxDb(module)
    client = influxdb.connect_to_influxdb()
    user = find_user(module, client, user_name)

    changed = False

    if state == 'present':
        if user:
            if not check_user_password(module, client, user_name, user_password) and user_password is not None:
                set_user_password(module, client, user_name, user_password)
                changed = True

            try:
                if admin and not user['admin']:
                    client.grant_admin_privileges(user_name)
                    changed = True
                elif not admin and user['admin']:
                    client.revoke_admin_privileges(user_name)
                    changed = True
            except influx.exceptions.InfluxDBClientError as e:
                module.fail_json(msg=to_native(e))

        else:
            user_password = user_password or ''
            create_user(module, client, user_name, user_password, admin)
            changed = True

        if grants:
            if set_user_grants(module, client, user_name, grants):
                changed = True
        else:
            if set_user_grants(module, client, user_name, []):
                changed = True

        module.exit_json(changed=changed)

    if state == 'absent':
        if user:
            drop_user(module, client, user_name)
        else:
            module.exit_json(changed=False)


if __name__ == '__main__':
    main()

